indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisorStateManager.java

```diff
@@ -0,0 +1,334 @@
+  private final CircularBuffer<State> stateHistory; // From previous runs
+  private final ExceptionEventStore eventStore;
+
+  public SeekableStreamSupervisorStateManager(
+      State initialState,
+      SeekableStreamSupervisorConfig supervisorConfig
+  )
+  {
+    this.supervisorState = initialState;
+    this.healthinessThreshold = supervisorConfig.getHealthinessThreshold();
+    this.unhealthinessThreshold = supervisorConfig.getUnhealthinessThreshold();
+    this.healthinessTaskThreshold = supervisorConfig.getTaskHealthinessThreshold();
+    this.unhealthinessTaskThreshold = supervisorConfig.getTaskUnhealthinessThreshold();
+    Preconditions.checkArgument(
+        supervisorConfig.getMaxStoredExceptionEvents() >=
+        Math.max(healthinessThreshold, unhealthinessThreshold),
+        "numExceptionEventsToStore must be greater than or equal to both "
+        + "healthinessThreshold and unhealthinessThreshold"
+    );
+    this.eventStore = new ExceptionEventStore(
+        supervisorConfig.getMaxStoredExceptionEvents(),
+        supervisorConfig.isStoringStackTraces()
+    );
+    this.completedTaskHistory = new CircularBuffer<>(Math.max(healthinessTaskThreshold, unhealthinessTaskThreshold));
+    this.stateHistory = new CircularBuffer<>(Math.max(healthinessThreshold, unhealthinessThreshold));
+  }
+
+  public Optional<State> setStateIfNoSuccessfulRunYet(State newState)
+  {
+    if (!atLeastOneSuccessfulRun) {
+      return Optional.of(setState(newState));
+    }
+    return Optional.absent();
+  }
+
+  public State setState(State newState)
+  {
+    if (newState.equals(State.SUSPENDED)) {
+      atLeastOneSuccessfulRun = false; // We want the startup states again after being suspended
+    }
+    this.supervisorState = newState;
+    return newState;
+  }
+
+  public void storeThrowableEvent(Throwable t)
+  {
+    if (t instanceof PossiblyTransientStreamException && atLeastOneSuccessfulRun) {
+      t = new TransientStreamException(t);
+    }
+
+    eventStore.storeThrowable(t);
+    currentRunSuccessful = false;
+  }
+
+  public void storeCompletedTaskState(TaskState state)
+  {
+    completedTaskHistory.add(state);
+  }
+
+  public void markRunFinishedAndEvaluateHealth()
+  {
+    if (currentRunSuccessful) {
+      atLeastOneSuccessfulRun = true;
+    }
+
+    State currentRunState = State.RUNNING;
+
+    for (Map.Entry<Class, Queue<ExceptionEvent>> events : eventStore.getNonTransientRecentEvents().entrySet()) {
+      if (events.getValue().size() >= unhealthinessThreshold) {
+        if (events.getKey().equals(NonTransientStreamException.class)) {
+          currentRunState = getHigherPriorityState(currentRunState, State.UNABLE_TO_CONNECT_TO_STREAM);
+        } else if (events.getKey().equals(TransientStreamException.class) ||
+                   events.getKey().equals(PossiblyTransientStreamException.class)) {
+          currentRunState = getHigherPriorityState(currentRunState, State.LOST_CONTACT_WITH_STREAM);
```

Comment:
Similarly, is it possible that you've never successfully connected to the stream, but you've only gotten PossiblyTransient or Transient exceptions?

Reply:
Yes.  I was debating whether to give the supervisor the higher severity `UNABLE_TO_CONNECT_TO_STREAM` state in that case.  I opted against doing so for the reason that I believe that it's better to incorrectly label a non-transient issue as transient than to incorrectly label a transient issue as non-transient (assuming someone's pager is hooked up to this endpoint).  Do you think there's value in subdividing the `UNABLE_TO_CONNECT_TO_STREAM` state into `UNABLE_TO_CONNECT_TO_STREAM_NON_TRANSIENT` and `UNABLE_TO_CONNECT_TO_STREAM_TRANSIENT` states.